### PR TITLE
RD-1450 Endpoint for creating Nodes

### PR DESCRIPTION
**Pull request by tehasdf**
Add an endpoint for batch-creating Nodes.

This is going to be useful for the "async deployment creation" track, where
we move creating nodes & node-instances out to a workflow. Then the workflow
of course needs a way to create those nodes.

### CODE REVIEW

File: rest-service/manager_rest/rest/resources_v3_1/nodes.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            max_number_of_instances=scalable.get('max_instances') or 1,
+            host_id=raw_node.get('host_id'),
+            properties=raw_node.get('properties') or {},
+            operations=raw_node.get('operations') or {},
+            plugins=raw_node.get('plugins') or {},
+            plugins_to_install=raw_node.get('plugins_to_install'),
+            relationships=self._prepare_node_relationships(raw_node)
+        )
+
+    def _prepare_node_relationships(self, raw_node):
```
Comment (tehasdf) [inline rest-service/manager_rest/rest/resources_v3_1/nodes.py:61]: this is just copied from resource_manager, but I'll leave it in there as well, until RD-1360 is done.

**Comment by tehasdf** [comment]
needs cloudify-cosmo/cloudify-common#667

File: rest-service/manager_rest/rest/resources_v3_1/nodes.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            deployment_id = self._deployment_id_from_nodes(raw_nodes)
+            deployment = sm.get(models.Deployment, deployment_id)
+            for raw_node in raw_nodes:
+                node = self._node_from_raw_node(raw_node)
+                node.set_deployment(deployment)
+                sm.put(node)
+                created.append(node)
+        return created, 200
+
+    def _deployment_id_from_nodes(self, raw_nodes):
```
Comment (geokala) [inline rest-service/manager_rest/rest/resources_v3_1/nodes.py:28]: Isn't this roughly equivalent to just doing raw_nodes[0]['deployment_id'], but with an effective extra implicit assertion that all raw nodes contain a deployment ID?

Comment (tehasdf) [reply rest-service/manager_rest/rest/resources_v3_1/nodes.py:28]: it is, you think there's a better way to do this? i think i want to have the check in some form

Comment (geokala) [reply rest-service/manager_rest/rest/resources_v3_1/nodes.py:28]: I think making it more explicit might be better, e.g.
deployment_id = raw_nodes[0]['deployment_id']
for node in raw_nodes:
-indent-eaten-by-github-    if node['deployment_id'] != deployment_id: raise BadThingHappened('I feel sad')
return deployment_id

File: rest-service/manager_rest/test/endpoints/test_nodes.py [hunk 3]
```diff
@@ -427,14 +427,24 @@ def test_node_and_node_instance_properties(self):
 
 
 class NodesCreateTest(base_test.BaseServerTestCase):
-    def test_create_single(self):
```
Comment (geokala) [inline rest-service/manager_rest/test/endpoints/test_nodes.py]: Do we not want to test creating a single node as well?

Comment (tehasdf) [reply rest-service/manager_rest/test/endpoints/test_nodes.py]: there isn't a way to create a single node, this only implements the "create batch" endpoint, which always takes a list. A test for 1-item list isn't meaningfully different I don't think?

File: rest-service/manager_rest/rest/resources_v3_1/nodes.py (hunk 2 above)
Comment (geokala) [reply rest-service/manager_rest/rest/resources_v3_1/nodes.py:28]: OK, I see what you did with this later- that makes it make more sense, so ignore this.

File: rest-service/manager_rest/test/endpoints/test_nodes.py [hunk 4]
```diff
@@ -448,3 +448,21 @@ def test_create_nodes(self):
         node1 = node1[0]
         assert len(node2) == 1
         assert node1.deployment_id == 'dep1'
+
+    def test_create_different_deployments(self):
+        self.put_blueprint()
+        self.client.deployments.create('blueprint', 'dep1')
+        self.client.deployments.create('blueprint', 'dep2')
+        with self.assertRaisesRegexp(CloudifyClientError, 'deployment'):
```
Comment (geokala) [inline rest-service/manager_rest/test/endpoints/test_nodes.py:456]: Maybe assert that it contains 'same deployment'?

File: rest-service/manager_rest/rest/resources_v3_1/nodes.py (hunk 2 above)
Comment (tehasdf) [reply rest-service/manager_rest/rest/resources_v3_1/nodes.py:28]: that makes for a worse error message, but OK, it's faster

**Review by geokala — APPROVED** [review]

**Review by geokala — APPROVED** [review]

**Pull request merged**